pallets/xyk/src/invariants.rs

```diff
@@ -0,0 +1,445 @@
+	desc: &str,
+) {
+	let new_s = U256::from(new_state.0) * U256::from(new_state.1);
+	let s1 = new_s.integer_sqrt();
+
+	let old_s = U256::from(old_state.0) * U256::from(old_state.1);
+	let s2 = old_s.integer_sqrt();
+
+	assert!(new_s >= old_s, "Invariant decreased for {}", desc);
+
+	let s1_u128 = Balance::try_from(s1).unwrap();
+	let s2_u128 = Balance::try_from(s2).unwrap();
+
+	let invariant = FixedU128::from((s1_u128, ONE)) / FixedU128::from((s2_u128, ONE));
+	assert_eq_approx!(invariant, FixedU128::from(1u128), tolerance, desc);
+}
+
+proptest! {
+	#![proptest_config(ProptestConfig::with_cases(1000))]
+	#[test]
+	fn add_liquidity(initial_liquidity in asset_reserve(),
+		added_liquidity in asset_reserve(),
+	) {
+		let asset_a = HDX;
+		let asset_b = DOT;
+
+		ExtBuilder::default()
+			.with_exchange_fee((0, 0))
+			.with_accounts(vec![
+				(ALICE, asset_a,initial_liquidity),
+				(ALICE, asset_b,initial_liquidity * 1000),
+				(BOB, asset_a, added_liquidity),
+				(BOB, asset_b, added_liquidity * 1_000_000),
+			])
+			.build()
+			.execute_with(|| {
+				assert_ok!(XYK::create_pool(
+					Origin::signed(ALICE),
+					asset_a,
+					asset_b,
+					initial_liquidity,
+					Price::from_float(0.6544)
+				));
+
+				let pool_account = XYK::get_pair_id(AssetPair {
+					asset_in: asset_a,
+					asset_out: asset_b,
+				});
+				let share_token = XYK::share_token(pool_account);
+
+				let pool_balance_a = Currency::free_balance(asset_a, &pool_account);
+				let pool_balance_b = Currency::free_balance(asset_b, &pool_account);
+
+				let bob_balance_a = Currency::free_balance(asset_a, &BOB);
+				let bob_balance_b = Currency::free_balance(asset_b, &BOB);
+
+				assert_ok!(XYK::add_liquidity(
+					Origin::signed(BOB),
+					asset_a,
+					asset_b,
+					added_liquidity,
+					added_liquidity * 1_000_000, // do not care about the limit here
+				));
+
+				let new_pool_balance_a = Currency::free_balance(asset_a, &pool_account);
+				let new_pool_balance_b = Currency::free_balance(asset_b, &pool_account);
+
+				let new_bob_balance_a = Currency::free_balance(asset_a, &BOB);
+				let new_bob_balance_b = Currency::free_balance(asset_b, &BOB);
+
+				let bob_shares = Currency::free_balance(share_token, &BOB);
+
+				let p0 = FixedU128::from((pool_balance_a, pool_balance_b));
+				let p1 = FixedU128::from((new_pool_balance_a, new_pool_balance_b));
+
+				// Price should not change
+				assert_eq_approx!(
+					p0,
+					p1,
+					FixedU128::from_float(0.0000000001),
+					"Price has changed after add liquidity"
+				);
+
+				let issuance = Currency::total_issuance(share_token);
+
+				let s = U256::from(issuance);
+				let delta_s = U256::from(bob_shares);
+				let delta_x = U256::from(bob_balance_a - new_bob_balance_a);
+				let delta_y = U256::from(bob_balance_b - new_bob_balance_b);
+				let x = U256::from(pool_balance_a);
+				let y = U256::from(pool_balance_b);
+
+				let l = delta_s * x;
+				let r = s * delta_x;
+
+				assert!(l <= r);
```

Comment:
It may be good to also add a check here that assert_eq_approx!(l, r, ...), for both the check with x and check with y.

These inequality checks assert that any discrepancy betweten l and r is in favor of the pool, but we also want l and r to be equal (up to rounding errors).